src/lightning/data/cache/worker.py

```diff
@@ -0,0 +1,331 @@
+                r = index_queue.get(timeout=MP_STATUS_CHECK_INTERVAL)
+            except queue.Empty:
+                continue
+            if isinstance(r, _ResumeIteration):
+                # Acknowledge the main process
+                data_queue.put((r, None))
+                iteration_end = False
+
+                if isinstance(dataset, IterDataPipe):
+                    assert r.seed is not None
+                    shared_rng.manual_seed(r.seed)
+                    dataset = apply_random_seed(dataset, shared_rng)
+
+                # Recreate the fetcher for worker-reuse policy
+                fetcher = _DatasetKind.create_fetcher(dataset_kind, dataset, auto_collation, collate_fn, drop_last)
+                continue
+
+            if r is None:
+                # Received the final signal
+                assert done_event.is_set() or iteration_end
+                break
+
+            if done_event.is_set() or iteration_end:
+                # `done_event` is set. But I haven't received the final signal
+                # (None) yet. I will keep continuing until get it, and skip the
+                # processing steps.
+                continue
+            idx, index = r
+            data: Union[_IterableDatasetStopIteration, ExceptionWrapper]
+            if init_exception is not None:
+                data = init_exception
+                init_exception = None
+            else:
+                try:
+                    data = fetcher.fetch(index)
+                except Exception as e:
+                    if isinstance(e, StopIteration) and dataset_kind == _DatasetKind.Iterable:
+                        data = _IterableDatasetStopIteration(worker_id)
+                        # Set `iteration_end`
+                        #   (1) to save future `next(...)` calls, and
+                        #   (2) to avoid sending multiple `_IterableDatasetStopIteration`s.
+                        iteration_end = True
+                    else:
+                        # It is important that we don't store exc_info in a variable.
+                        # `ExceptionWrapper` does the correct thing.
+                        # See NOTE [ Python Traceback Reference Cycle Problem ]
+                        data = ExceptionWrapper(where=f"in DataLoader worker process {worker_id}")
+            data_queue.put((idx, data))
+            del data, idx, index, r  # save memory
+    except KeyboardInterrupt:
+        # Main process will raise KeyboardInterrupt anyways.
+        pass
+    if done_event.is_set():
+        ####### ADDTIONATIONAL CODE #######
+
+        from lightning.data.cache import Cache
+
+        # required to ensure the cache is persisted
+        if dataset_kind == _DatasetKind.Map:
+            for v in fetcher.dataset.__dict__.values():
+                if isinstance(v, Cache):
+                    v.done()
+
+        ####### ADDTIONATIONAL CODE #######
+
+        data_queue.cancel_join_thread()
+        data_queue.close()
```

Comment:
why?

Reply:
The entire worker.py file is taken from PyTorch. It is written at the top. Don't need to review this code.

The `ADDTIONATIONAL CODE` is the part to ensure we are writing the index file.

Reply:
when it's taken from pytorch, why can't we just import from there?

Reply:
Ok, found an alternative:

```
class WorkerLoop:
    def __call__(self, dataset_kind, *args, **kwargs):
        from torch.utils.data import _DatasetKind
        from torch.utils.data._utils import worker

        from lightning.data.cache.cache import Cache

        reloaded_worker = reload(worker)

        create_fetcher = _DatasetKind.create_fetcher

        fetcher = None

        def create_fetcher_fn(*args, **kwargs):
            nonlocal fetcher
            fetcher = create_fetcher(*args, **kwargs)
            return fetcher

        _DatasetKind.create_fetcher = create_fetcher_fn

        reloaded_worker._worker_loop(dataset_kind, *args, **kwargs)

        if dataset_kind == _DatasetKind.Map:
            for v in fetcher.dataset.__dict__.values():
                if isinstance(v, Cache):
                    v.done()


class _MultiProcessingDataLoaderIterPatch(_MultiProcessingDataLoaderIter):
    def __init__(self, loader):
        # Patch PyTorch worker loop to call the `cache.done()` method.
        from torch.utils.data._utils import worker

        worker._worker_loop = WorkerLoop()
        super().__init__(loader)
```